src/testing/cmds/harness/test-collector/main.go

```diff
@@ -0,0 +1,182 @@
+	if exclusiveTags && tags != "" {
+		// set difference to get ONLY tagged tests
+		var err error
+		testIdsUntagged, err := testNames(ctx, pkg, "") // collect for set difference
+		if err != nil {
+			return errors.EnsureStack(err)
+		}
+		testIds = subtractTestSet(testIdsTagged, testIdsUntagged)
+	} else {
+		testIds = testIdsTagged
+	}
+	log.Info(ctx, "tests and packages collected", zap.Any("tests", testIds))
+	err = outputToFile(fileName, testIds)
+	return err
+}
+
+// get tests that are tagged, but not in the untagged list since that is inclusive
+func subtractTestSet(testIdsTagged map[string][]string, testIdsUntagged map[string][]string) map[string][]string {
+	resultTests := map[string][]string{}
+	for pkg, testsTagged := range testIdsTagged {
+		testsUntagged, ok := testIdsUntagged[pkg]
+		for _, testNameTagged := range testsTagged {
+			if !ok || !slice.ContainsString(testsUntagged, testNameTagged, func(s string) string { return s }) {
+				if _, ok := resultTests[pkg]; !ok {
+					resultTests[pkg] = []string{testNameTagged}
+				} else {
+					resultTests[pkg] = append(resultTests[pkg], testNameTagged)
+				}
+			}
+		}
+
+	}
+	return resultTests
+}
+
+func testNames(ctx context.Context, pkg string, addtlCmdArgs ...string) (map[string][]string, error) {
+	findTestArgs := append([]string{"test", pkg, "-json", "-list=."}, addtlCmdArgs...)
+	cmd := exec.Command("go", findTestArgs...)
+	cmdReader, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, "GOMAXPROCS=16") // This prevents the command from running wild eating up processes in the pipelines
+	stderr, err := cmd.StderrPipe()
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	scanner := bufio.NewScanner(cmdReader)
+	err = cmd.Start()
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	var testNames = map[string][]string{}
+	for scanner.Scan() {
+		testInfo := &testOutput{}
+		raw := scanner.Bytes()
+		if !bytes.HasPrefix(raw, []byte("{")) {
+			continue // dependency download junk got shared
+		}
+		err := json.Unmarshal(raw, testInfo)
+		if err != nil {
+			return nil, errors.Wrapf(err, "parsing json: %s", string(raw))
+		}
+		if testInfo.Action == "output" {
+			output := strings.Trim(testInfo.Output, "\n ")
+			if output != "" && !strings.HasPrefix(output, "Benchmark") &&
+				!strings.HasPrefix(output, "ExampleAPIClient_") &&
+				!strings.HasPrefix(output, "? ") &&
+				!strings.HasPrefix(output, "ok ") {
+				if _, ok := testNames[testInfo.Package]; !ok {
+					testNames[testInfo.Package] = []string{output}
+				} else {
+					testNames[testInfo.Package] = append(testNames[testInfo.Package], output)
+				}
+			}
+
+		}
+
+	}
+	stderrBytes, err := io.ReadAll(stderr)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	errorText := string(stderrBytes)
+	if errorText != "" {
+		log.Error(ctx, string(errorText))
+	}
+
+	err = cmd.Wait()
```

Comment:
It's probably fine to just assume that if the command exited with status 0 that everything is OK.  No need to ensure that stderr didn't get written to.